SmartTutor/Communication/SmartTutorMessageConsumer.cs

```diff
@@ -0,0 +1,78 @@
+﻿using Newtonsoft.Json;
+using RabbitMQ.Client;
+using RabbitMQ.Client.Events;
+using SmellDetector.SmellModel.Reports;
+using System;
+using System.Linq;
+using System.Text;
+
+namespace SmartTutor.Communucation
+{
+
+    public class SmartTutorMessageConsumer
+    {
+        public string NodeName { get; set; }
+        public string QueueName { get; set; }
+        public string ExchangeName { get; set; }
+        public IConnection Connection { get; set; }
+        public IModel Channel { get; set; }
+
+        public SmartTutorMessageConsumer()
+        {
+            ConfigureInitialStates();
+            CreateConnection();
+            Channel = Connection.CreateModel();
+            DeclareQueue();
+            ConsumeMessage(DecodeMessage());
+        }
+
+        private void ConfigureInitialStates()
+        {
+            NodeName = "localhost";
+            QueueName = "IssueReports";
+            ExchangeName = "";
+        }
+
+        private void CreateConnection()
+        {
+            var connectionFactory = new ConnectionFactory() { HostName = NodeName };
+            Connection = connectionFactory.CreateConnection();
+        }
+
+        private void DeclareQueue()
+        {
+            Channel.QueueDeclare(queue: QueueName,
+                                                 durable: false,
+                                                 exclusive: false,
+                                                 autoDelete: false,
+                                                 arguments: null);
+        }
+
+        private EventingBasicConsumer DecodeMessage()
+        {
+            var consumer = new EventingBasicConsumer(Channel);
+            consumer.Received += (model, deliveryArguments) =>
+            {
+                var body = deliveryArguments.Body.ToArray();
+                var jsonMessage = Encoding.UTF8.GetString(body);
+                SmellDetectionReport reportMessage = new SmellDetectionReport();
```

Comment:
When we introduce the CommunicationUtilities project, we need to think about how to define the generic consumer and producer classes. This should be done by examining the code and seeing which parts are duplicate and which aren't.

Currently I think that most of the code can be centralized, and the differences (including this callback function) can be passed through constructor parameters. We will discuss this on Friday.

Reply:
I also noticed that, but not worked on it. We definitely could make some improvement on this.